torch/nn/utils/attention.py

```diff
@@ -0,0 +1,253 @@
+""" Define Base class as well as some crowd favorites """
+from abc import ABC, abstractmethod
+from enum import IntEnum
+from typing import Optional, Union
+from warnings import warn
+
+import torch
+from torch.backends.cuda import (
+    can_use_efficient_attention,
+    can_use_flash_attention,
+    SDPAParams,
+)
+from torch.nn.functional import scaled_dot_product_attention
+from torch.utils import _pytree as pytree
+
+
+__all__ = ["AttnBias", "TensorBias", "CausalVariant", "CausalBias"]
+
+
+def _input_requires_grad(*tensors: torch.Tensor) -> bool:
+    """Returns True if any of the tensors requires grad"""
+    return any(t.requires_grad for t in tensors)
+
+
+def _materialize_if_needed(
+    bias: "AttnBias", device: Optional[torch.device] = None
+) -> Union[torch.Tensor, "AttnBias"]:
+    if bias.needs_materialization():
+        return bias.materialize(device)
+    return bias
+
+
+def _postprocess_flash_output(inpt_tensor: torch.Tensor, og_size: int):
+    """Handles the unpad of the last dimension"""
+    if inpt_tensor.size(-1) != og_size:
+        return inpt_tensor[..., :og_size]
+    return inpt_tensor
+
+
+class AttnBias(ABC):
+    """Abstract base class for attention biases"""
+
+    @abstractmethod
+    def materialize(self, device: Optional[torch.device] = None) -> torch.Tensor:
+        raise NotImplementedError("This is an abstract base class")
+
+    @abstractmethod
+    def needs_materialization(self) -> bool:
+        raise NotImplementedError("This is an abstract base class")
+
+    @staticmethod
+    @abstractmethod
+    def dispatch(
+        query: torch.Tensor,
+        key: torch.Tensor,
+        value: torch.Tensor,
+        attn_mask,
+        dropout_p: float,
+        is_causal: bool,
+        scale: Optional[float],
+    ) -> torch.Tensor:
+        raise NotImplementedError("This is an abstract base class")
```

Comment:
hm it doesn't seem like this ABC is giving you much if you have to implement this function for your subclass AND manually set up a `Subclass.__torch_function__()` that calls it. It'd be nice if the base handled the `__torch_function__()` part

Reply:
ahhh good idea!

Reply:
And yeah this ABC is more of a helper "scaffolding" that I think will be helpful for people wanting to implement their own masks